### PR TITLE
Adding NMSv4 GPU implementation

### PR DESCRIPTION
**Pull request by samikama**
This PR adds GPU implementation of NonMaxSuppresionOpV4.

### CODE REVIEW

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 1]
```diff
@@ -529,16 +534,31 @@ Status DoNMS(OpKernelContext* context, const Tensor& boxes,
   }
   Tensor* output_indices = nullptr;
   int num_outputs = std::min(num_to_keep, output_size);  // no padding!
-  TF_RETURN_IF_ERROR(
-      context->allocate_output(0, TensorShape({num_outputs}), &output_indices));
-  if (num_outputs == 0) return Status::OK();
+  if (pad_to_max_output && num_outputs!=max_output_size) {
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Please remove line 427 (alias max_output_size). It's confusing to have both output_size and max_output_size meaning the same thing.

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 2]
```diff
@@ -646,25 +668,103 @@ class NonMaxSuppressionV3GPUOp : public OpKernel {
...
+class NonMaxSuppressionV4GPUOp : public OpKernel {
+ public:
+  explicit NonMaxSuppressionV4GPUOp(OpKernelConstruction* context)
+      : OpKernel(context) {
+    OP_REQUIRES_OK(context, context->GetAttr("pad_to_max_output_size",
+                                             &pad_to_max_output_size_));
+  }
+
+  void Compute(OpKernelContext* context) override {
+    // boxes: [num_boxes, 4]
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Can we add a helper method for these input validation logic as well? They're almost the same for v2/v3/v4 op.

**Review by aaroey — COMMENTED** [review]
Thanks for the PR! Just some minor comments then we're good to go.

**Review by aaroey — APPROVED** [review]

**Comment by qinyao-he** [comment]
@samikama Is there any plan to add CombinedNonMaxSuppression GPU version?

**Comment by samikama** [comment]
@qinyao-he check #34852

**Pull request merged**